src/adapter/src/explain/optimizer_trace.rs

```diff
@@ -56,33 +55,37 @@ impl OptimizerTrace {
         if broken {
             let subscriber = DelegateSubscriber::default()
                 // Collect `explain_plan` types that are not used in the regular explain
                 // path, but are useful when instrumenting code for debugging purpuses.
-                .with(PlanTrace::<String>::new(path))
-                .with(PlanTrace::<HirScalarExpr>::new(path))
-                .with(PlanTrace::<MirScalarExpr>::new(path))
+                .with(PlanTrace::<String>::new(filter))
+                .with(PlanTrace::<HirScalarExpr>::new(filter))
+                .with(PlanTrace::<MirScalarExpr>::new(filter))
                 // Collect `explain_plan` types that are used in the regular explain path.
-                .with(PlanTrace::<HirRelationExpr>::new(path))
-                .with(PlanTrace::<MirRelationExpr>::new(path))
-                .with(PlanTrace::<DataflowDescription<OptimizedMirRelationExpr>>::new(path))
-                .with(PlanTrace::<DataflowDescription<Plan>>::new(path));
+                .with(PlanTrace::<HirRelationExpr>::new(filter))
+                .with(PlanTrace::<MirRelationExpr>::new(filter))
+                .with(PlanTrace::<DataflowDescription<OptimizedMirRelationExpr>>::new(filter))
+                .with(PlanTrace::<DataflowDescription<Plan>>::new(filter))
+                // Don't filter for FastPathPlan entries (there can be at most one).
+                .with(PlanTrace::<FastPathPlan>::new(None));
 
             OptimizerTrace(dispatcher::Dispatch::new(subscriber))
         } else {
             let subscriber = tracing_subscriber::registry()
                 // Collect `explain_plan` types that are not used in the regular explain
                 // path, but are useful when instrumenting code for debugging purpuses.
-                .with(PlanTrace::<String>::new(path))
-                .with(PlanTrace::<HirScalarExpr>::new(path))
-                .with(PlanTrace::<MirScalarExpr>::new(path))
+                .with(PlanTrace::<String>::new(filter))
+                .with(PlanTrace::<HirScalarExpr>::new(filter))
+                .with(PlanTrace::<MirScalarExpr>::new(filter))
                 // Collect `explain_plan` types that are used in the regular explain path.
-                .with(PlanTrace::<HirRelationExpr>::new(path))
-                .with(PlanTrace::<MirRelationExpr>::new(path))
-                .with(PlanTrace::<DataflowDescription<OptimizedMirRelationExpr>>::new(path))
-                .with(PlanTrace::<DataflowDescription<Plan>>::new(path));
+                .with(PlanTrace::<HirRelationExpr>::new(filter))
+                .with(PlanTrace::<MirRelationExpr>::new(filter))
+                .with(PlanTrace::<DataflowDescription<OptimizedMirRelationExpr>>::new(filter))
+                .with(PlanTrace::<DataflowDescription<Plan>>::new(filter))
+                // Don't filter for FastPathPlan entries (there can be at most one).
+                .with(PlanTrace::<FastPathPlan>::new(None));
```

Comment:
Unfortunately that's not so easy because some type-level stuff that causes the resulting type to be always dependent on the call stack. Probably can factor out this into a function, but I'm not sure if people won't attempt to murder me when they see it's type signature :(

Reply:
Ok, then it's not so important :)